### PR TITLE
Add non canonical block storage and retrieval

### PR DESCRIPTION
**Pull request by rolfyone**
- Get block by root will retrieve non canonical blocks

 - enabled via development cli arg `--Xdata-storage-non-canonical-blocks-enabled`

 Non canonical blocks can't be retrieved, so the client would need to actually see the non canonical block for it to be stored. Debug trace has been added to output to log file when a non canonical block is seen.

 partially addresses #2853

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: services/chainstorage/src/main/java/tech/pegasys/teku/services/chainstorage/StorageConfiguration.java [hunk 1]
```diff
@@ -59,6 +62,10 @@ public DatabaseVersion getDataStorageCreateDbVersion() {
     return dataStorageCreateDbVersion;
   }
 
+  public boolean isStoreNonCanonicalBlocks() {
```
Comment (ajsutton) [inline services/chainstorage/src/main/java/tech/pegasys/teku/services/chainstorage/StorageConfiguration.java]: Maybe
```suggestion
  public boolean isStoreNonCanonicalBlocksEnabled() {
```

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java [hunk 2]
```diff
@@ -616,6 +656,18 @@ public void pruneHotStateRoots(final List<Bytes32> stateRoots) {
...
+  public void addNonCanonicalBlock(final SignedBeaconBlock block) {
+    if (!storeNonCanonicalBlocks) {
+      return;
+    }
+    try (final FinalizedUpdater updater = finalizedDao.finalizedUpdater()) {
+      LOG.debug("Non canonical block {}", block.getRoot().toHexString());
+      updater.addNonCanonicalBlock(block);
+      updater.commit();
+    }
+  }
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java]: We don't want to use a new transaction for every block we store.  Suspect this method should be on `FinalizedUpdater` instead so it just adds them to the existing transaction and then they are committed with the rest of the changes.

File: teku/src/main/java/tech/pegasys/teku/cli/options/DataStorageOptions.java [hunk 3]
```diff
@@ -43,6 +43,14 @@
       hidden = true)
   private String createDbVersion = DatabaseVersion.DEFAULT_VERSION.getValue();
 
+  @CommandLine.Option(
+      names = {"--Xdata-storage-non-canonical-blocks-enabled"},
+      paramLabel = "<BOOLEAN>",
+      description = "Store non canonical blocks",
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/options/DataStorageOptions.java]:
```suggestion
      description = "Store non-canonical blocks",
```

**Review by l2ichshop2 — APPROVED** [review]

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java [hunk 4]
```diff
@@ -728,6 +770,37 @@ private void updateFinalizedData(
...
         throw new UnsupportedOperationException("Unhandled storage mode: " + stateStorageMode);
     }
+
+    if (storeNonCanonicalBlocks) {
+      storeNonCanonicalBlocks(
+          deletedHotBlocks.stream()
+              .filter(root -> !finalizedChildToParentMap.containsKey(root))
+              .map(this::getHotBlock)
+              .filter(Optional::isPresent)
+              .map(Optional::get)
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java]:
```suggestion
              .flatMap(root -> getHotBlock().stream())
```

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java [hunk 5]
```diff
@@ -728,6 +770,37 @@ private void updateFinalizedData(
...
+              .collect(Collectors.toSet()));
+    }
+  }
+
+  private void storeNonCanonicalBlocks(final Set<SignedBeaconBlock> nonCanonicalBlocks) {
+    if (!storeNonCanonicalBlocks) {
+      return;
+    }
+    checkNotNull(nonCanonicalBlocks, "Non-canonical block set is null");
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java]: I'm not sure we actually need this - it's a private method and given the way it's called can never be given a null.

**Review by ajsutton — APPROVED** [review]
LGTM.

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java (hunk 5 above)
Comment (rolfyone) [reply storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java]: I had just put it in for safety but i can remove it. I think i had it in before i changed structure

**Pull request merged**